File: src/hyper_markdown/scan.py
"""Masking scanner: find the constructs the format owns (HMD-0001 §1, §2).

The scanner does not parse CommonMark. It masks the regions where markup must
not be interpreted, then extracts constructs from what is left. Masking replaces
characters with spaces and preserves newlines, so every byte offset, line, and
column in the masked text still refers to the original source.
"""

from __future__ import annotations

import re

# A fence opener/closer: three or more backticks or tildes, optionally indented.
_FENCE_RE = re.compile(r"^(?P<indent>[ \t]*)(?P<fence>`{3,}|~{3,})(?P<info>.*)$")
_HTML_COMMENT_RE = re.compile(r"<!--.*?-->", re.DOTALL)
# A code span: a run of N backticks, then the shortest text, then N backticks.
_CODE_SPAN_RE = re.compile(r"(?P<ticks>`+)(?P<body>.+?)(?P=ticks)", re.DOTALL)

_ATX_HEADING_RE = re.compile(r"^[ \t]{0,3}(?P<hashes>#{1,6})[ \t]+(?P<text>.*?)[ \t]*$", re.MULTILINE)
_ANCHOR_RE = re.compile(r"[ \t]+\^(?P<id>[A-Za-z0-9][A-Za-z0-9_-]{0,63})[ \t]*$", re.MULTILINE)
# A wikilink target may not contain "[", "]" or a newline (HMD-0001 §2).
_LINK_RE = re.compile(r"(?P<bang>!?)\[\[(?P<inner>[^\[\]\n]*)\]\]")
_OPEN_RE = re.compile(r"!?\[\[")


def mask(text: str) -> str:
    """Return `text` with uninterpreted regions blanked out, offsets preserved.

    Masked: fenced code blocks, HTML comments, and inline code spans.

    Indented code blocks are deliberately NOT masked. This tree ships the
    `admonition` and `footnotes` extensions, under which a four-space indent
    marks callout bodies and footnote continuations rather than code; masking it
    would silently drop real links from ordinary prose.
    """
    chars = list(text)
    for start, end in _masked_regions(text):
        for i in range(start, end):
            if chars[i] != "\n":
                chars[i] = " "
    return "".join(chars)


def _masked_regions(text: str) -> list[tuple[int, int]]:
    regions = list(_fenced_regions(text))
    blanked = _blank(text, regions)

    # Comments and code spans are found in text that already has fences removed,
    # so a "<!--" inside a code block cannot open a comment.
    for m in _HTML_COMMENT_RE.finditer(blanked):
        regions.append((m.start(), m.end()))
    blanked = _blank(blanked, regions)

    for m in _CODE_SPAN_RE.finditer(blanked):
        if "\n\n" in m.group(0):  # a code span cannot span a blank line
            continue
        regions.append((m.start(), m.end()))
    return regions


def _blank(text: str, regions: list[tuple[int, int]]) -> str:
    chars = list(text)
    for start, end in regions:
        for i in range(start, end):
            if chars[i] != "\n":
                chars[i] = " "
    return "".join(chars)


def _fenced_regions(text: str):
    """Yield (start, end) for each fenced code block, closed or not."""
    open_fence: str | None = None
    open_start = 0
    offset = 0
    for line in text.splitlines(keepends=True):
        stripped = line.rstrip("\r\n")
        m = _FENCE_RE.match(stripped)
        if open_fence is None:
            if m and not m.group("info").strip().startswith(m.group("fence")[0]):
                open_fence = m.group("fence")
                open_start = offset
        else:
            # A closer is the same character, at least as long, with no info string.
            if (
                m
                and m.group("fence")[0] == open_fence[0]
                and len(m.group("fence")) >= len(open_fence)
                and not m.group("info").strip()
            ):
                yield (open_start, offset + len(line))
                open_fence = None
        offset += len(line)
    if open_fence is not None:  # unterminated fence runs to end of file
        yield (open_start, len(text))


def line_col(text: str, offset: int) -> tuple[int, int]:
    """1-indexed line and column for a byte offset."""
    line = text.count("\n", 0, offset) + 1
    last_nl = text.rfind("\n", 0, offset)
    return line, offset - last_nl


def find_headings(masked: str):
    """Yield (level, text, span_start, span_end) for each ATX heading."""
    for m in _ATX_HEADING_RE.finditer(masked):
        text = m.group("text")
        # Drop a closing hash run, then any trailing block anchor.
        text = re.sub(r"[ \t]+#+[ \t]*$", "", text)
        text = re.sub(r"[ \t]+\^[A-Za-z0-9][A-Za-z0-9_-]{0,63}[ \t]*$", "", text)
        yield len(m.group("hashes")), text.strip(), m.start(), m.end()


def find_anchors(masked: str):
    """Yield (block_id, span_start, span_end) for each trailing ``^id``."""
    for m in _ANCHOR_RE.finditer(masked):
        yield m.group("id"), m.start(), m.end()


def find_links(masked: str):
    """Yield (bang, inner, span_start, span_end) for each well-formed link."""
    for m in _LINK_RE.finditer(masked):
        yield m.group("bang") == "!", m.group("inner"), m.start(), m.end()


def find_unterminated(masked: str):
    """Yield the offset of every ``[[`` that does not open a complete link."""
    closed = {m.start() + (1 if m.group("bang") else 0) for m in _LINK_RE.finditer(masked)}
    for m in _OPEN_RE.finditer(masked):
        start = m.start() + (1 if m.group(0).startswith("!") else 0)
        if start not in closed:
            yield m.start()

File: tools/hmd-ts-core/src/scan.ts
/**
 * Masking scanner (HMD-0001 §1, §2) — a port of `src/hyper_markdown/scan.py`.
 *
 * The scanner does not parse CommonMark. It masks the regions where markup must
 * not be interpreted, then extracts constructs from what is left. Masking
 * replaces characters with spaces and preserves newlines, so every offset,
 * line, and column in the masked text still refers to the original source.
 *
 * HMD-0020 §3 gives markdown-it the rendering job. It does not get this one:
 * conformance with the canonical implementation is measured here, and the two
 * scanners disagree on exactly the case HMD-0001 §1 legislates — an indented
 * block is a callout body, not code.
 */

const FENCE_RE = /^([ \t]*)(`{3,}|~{3,})(.*)$/;
const HTML_COMMENT_RE = /<!--[\s\S]*?-->/g;
const CODE_SPAN_RE = /(`+)([\s\S]+?)\1/g;

const ATX_HEADING_RE = /^[ \t]{0,3}(#{1,6})[ \t]+(.*?)[ \t]*$/gm;
const ANCHOR_RE = /[ \t]+\^([A-Za-z0-9][A-Za-z0-9_-]{0,63})[ \t]*$/gm;
/** A wikilink target may not contain "[", "]" or a newline (HMD-0001 §2). */
const LINK_RE = /(!?)\[\[([^[\]\n]*)\]\]/g;
const OPEN_RE = /!?\[\[/g;

export type Region = readonly [number, number];

/** Split `text` into lines, keeping the line terminators. */
function linesWithEnds(text: string): string[] {
  const out: string[] = [];
  let start = 0;
  for (let i = 0; i < text.length; i += 1) {
    if (text[i] === "\n") {
      out.push(text.slice(start, i + 1));
      start = i + 1;
    }
  }
  if (start < text.length) out.push(text.slice(start));
  return out;
}

function stripEol(line: string): string {
  return line.replace(/\r?\n$/, "");
}

/** Yield [start, end) for each fenced code block, closed or not. */
function fencedRegions(text: string): Region[] {
  const regions: Region[] = [];
  let openFence: string | null = null;
  let openStart = 0;
  let offset = 0;

  for (const line of linesWithEnds(text)) {
    const stripped = stripEol(line);
    const m = FENCE_RE.exec(stripped);
    if (openFence === null) {
      if (m && !m[3]!.trim().startsWith(m[2]![0]!)) {
        openFence = m[2]!;
        openStart = offset;
      }
    } else if (
      m &&
      m[2]![0] === openFence[0] &&
      m[2]!.length >= openFence.length &&
      m[3]!.trim() === ""
    ) {
      regions.push([openStart, offset + line.length]);
      openFence = null;
    }
    offset += line.length;
  }

  // An unterminated fence runs to end of file.
  if (openFence !== null) regions.push([openStart, text.length]);
  return regions;
}

function blank(text: string, regions: readonly Region[]): string {
  // split("") keeps UTF-16 code units, so offsets stay aligned with the
  // indices RegExp reports. Array.from would split by code point and shift
  // every offset after the first astral character.
  const chars = text.split("");
  for (const [start, end] of regions) {
    for (let i = start; i < end && i < chars.length; i += 1) {
      if (chars[i] !== "\n") chars[i] = " ";
    }
  }
  return chars.join("");
}

function maskedRegions(text: string): Region[] {
  const regions: Region[] = [...fencedRegions(text)];
  let blanked = blank(text, regions);

  // Comments and code spans are found in text that already has fences removed,
  // so a "<!--" inside a code block cannot open a comment.
  HTML_COMMENT_RE.lastIndex = 0;
  for (const m of blanked.matchAll(HTML_COMMENT_RE)) {
    regions.push([m.index!, m.index! + m[0].length]);
  }
  blanked = blank(blanked, regions);

  CODE_SPAN_RE.lastIndex = 0;
  for (const m of blanked.matchAll(CODE_SPAN_RE)) {
    if (m[0].includes("\n\n")) continue; // a code span cannot span a blank line
    regions.push([m.index!, m.index! + m[0].length]);
  }
  return regions;
}

/**
 * Return `text` with uninterpreted regions blanked out, offsets preserved.
 *
 * Masked: fenced code blocks, HTML comments, and inline code spans.
 *
 * Indented code blocks are deliberately NOT masked. Under `admonition` and
 * `footnotes` a four-space indent marks a callout body or a footnote
 * continuation rather than code; masking it would silently drop real links
 * from ordinary prose.
 */
export function mask(text: string): string {
  return blank(text, maskedRegions(text));
}

/** An index from 0-indexed line number to that line's start offset. */
export function lineStarts(text: string): number[] {
  const starts = [0];
  for (let i = 0; i < text.length; i += 1) {
    if (text[i] === "\n") starts.push(i + 1);
  }
  return starts;
}

/**
 * Offset -> position lookup, built once per parse.
 *
 * A linear scan per construct is O(n·m) on a document with many links, which
 * is exactly the shape of the documents this format encourages.
 */
export class PositionIndex {
  private readonly starts: number[];

  constructor(text: string) {
    this.starts = lineStarts(text);
  }

  /** 1-indexed line and column for an offset. */
  at(offset: number): { line: number; column: number } {
    let lo = 0;
    let hi = this.starts.length - 1;
    while (lo < hi) {
      const mid = (lo + hi + 1) >> 1;
      if (this.starts[mid]! <= offset) lo = mid;
      else hi = mid - 1;
    }
    return { line: lo + 1, column: offset - this.starts[lo]! + 1 };
  }

  /** Start offset of a 1-indexed line. */
  offsetOfLine(line: number): number {
    return this.starts[Math.min(Math.max(line, 1), this.starts.length) - 1]!;
  }

  get lineCount(): number {
    return this.starts.length;
  }
}

/** 1-indexed line and column for an offset. Prefer `PositionIndex` in loops. */
export function lineCol(text: string, offset: number): { line: number; column: number } {
  return new PositionIndex(text).at(offset);
}

export interface RawHeading {
  level: number;
  text: string;
  start: number;
  end: number;
}

const CLOSING_HASHES_RE = /[ \t]+#+[ \t]*$/;
const TRAILING_ANCHOR_RE = /[ \t]+\^[A-Za-z0-9][A-Za-z0-9_-]{0,63}[ \t]*$/;

export function findHeadings(masked: string): RawHeading[] {
  const out: RawHeading[] = [];
  ATX_HEADING_RE.lastIndex = 0;
  for (const m of masked.matchAll(ATX_HEADING_RE)) {
    let text = m[2]!;
    // Drop a closing hash run, then any trailing block anchor.
    text = text.replace(CLOSING_HASHES_RE, "");
    text = text.replace(TRAILING_ANCHOR_RE, "");
    out.push({
      level: m[1]!.length,
      text: text.trim(),
      start: m.index!,
      end: m.index! + m[0].length,
    });
  }
  return out;
}

export interface RawAnchor {
  blockId: string;
  start: number;
  end: number;
}

export function findAnchors(masked: string): RawAnchor[] {
  const out: RawAnchor[] = [];
  ANCHOR_RE.lastIndex = 0;
  for (const m of masked.matchAll(ANCHOR_RE)) {
    out.push({ blockId: m[1]!, start: m.index!, end: m.index! + m[0].length });
  }
  return out;
}

export interface RawLink {
  isEmbed: boolean;
  inner: string;
  start: number;
  end: number;
}

export function findLinks(masked: string): RawLink[] {
  const out: RawLink[] = [];
  LINK_RE.lastIndex = 0;
  for (const m of masked.matchAll(LINK_RE)) {
    out.push({
      isEmbed: m[1] === "!",
      inner: m[2]!,
      start: m.index!,
      end: m.index! + m[0].length,
    });
  }
  return out;
}

/** Offsets of every `[[` that does not open a complete link. */
export function findUnterminated(masked: string): number[] {
  const closed = new Set<number>();
  LINK_RE.lastIndex = 0;
  for (const m of masked.matchAll(LINK_RE)) {
    closed.add(m.index! + (m[1] === "!" ? 1 : 0));
  }

  const out: number[] = [];
  OPEN_RE.lastIndex = 0;
  for (const m of masked.matchAll(OPEN_RE)) {
    const start = m.index! + (m[0].startsWith("!") ? 1 : 0);
    if (!closed.has(start)) out.push(m.index!);
  }
  return out;
}
